code/datums/components/shell.dm

```diff
@@ -141,7 +145,9 @@
 
 /datum/component/shell/proc/on_circuit_add_component(datum/source, obj/item/circuit_component/added_comp)
 	SIGNAL_HANDLER
-	return COMPONENT_CANCEL_ADD_COMPONENT
+
+	if (!added_comp.insertable_through_shells)
```

Comment:
for this, it may just be better to change the signal to avoid adding circuit components from users, and just not generally denying circuits from being added in general so that future code doesn't need to look at adding insertable_through_shells if they want to forcefully add a component on.

This'd require an extra signal that's sent when a user tries to add a component.